### PR TITLE
ENH: add np.nan funcs to _cython_table

### LINKED ISSUE #19629: series.agg(np.nansum) etc. returns a  series, scalar expected
```python
>>> pd.Series([1,2,3,4]).agg(np.sum)
10
>>> pd.Series([1,2,3,4]).agg(np.nansum)
0    1
1    2
2    3
3    4
dtype: int64
```
I would expect the return values of ``.agg(np.nansum)`` and ``agg(np.sum)`` to be the same. 

the same issue is with np.nanmean, np.nanmax etc.

#### Problem description

The underlying issue is that ``SelectionMixIn._cython_table`` has records for ``np.sum``, ``np.mean`` etc, but not ``np.nansum``, ``np.nanmean`` etc.

From a quick look at it, it seems that the issue is easily fixable, by adding the ``np.nan*`` function to cython_table.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 7590b8723516a9d595c287067fa639fbb9e19ec1
python: 3.6.3.final.0
python-bits: 32
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.22.0.dev0+710.g7590b87.dirty
pytest: 3.3.1
pip: 9.0.1
setuptools: 38.2.5
Cython: 0.26.1
numpy: 1.13.3
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.0
openpyxl: 2.4.9
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0b10
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #21134: BUG: df.agg(sum, axis=1) gives wrong result when Nan value is in frame
Using the built-in ``sum`` function gives correct result in ``df.agg(sum, axis=0)``, but wrong result in ``df.agg(sum, axis=1)``.

```python
>>> df = pd.DataFrame([[np.nan, 2], [3, 4]])
>>> df.agg(sum, axis=0)
0    3.0
1    6.0
dtype: float64
>>> df.agg(sum, axis=1)
0    NaN
1    7.0
dtype: float64
```

The ``NaN`` in the last example should be 2.0. 

Also, operation using the builtin ``sum`` in ``agg`` with ``axis=1`` are very slow:

```python
>>> n = 1_000
>>> df = pd.DataFrame({'a': range(n), 'b': range(n)})
>>> %timeit df.agg(sum, axis=1)
16.5 ms ± 25.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
>>> %timeit df.T.agg(sum)  # correct result *and* faster
312 µs ± 3.17 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

#### Problem description

Currently, ``df.agg(func, axis=1)`` defers to ``df.apply(func, axis=1)``. This is not done for ``axis=0``, and the operation may therefore give unexpected results and slow the operation down (because ``df.apply`` can be very slow).

#### Expected Output

The expected output is:

```python
>>> df.agg(sum, axis=1)
0    2.0
1    7.0
dtype: float64
```

#### Solution proposal

I'm thinking about putting in ``df.T.agg(func, axis=0)`` rather than ``df.apply(func, axis=1)`` in a few strategic places. This should ensure both getting correct results and faster operations. will report back if this succeeds.

### PR DESCRIPTION
**Pull request by topper-123**
closes #19629 
closes #21134
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This started as a copy of #19670 by @AaronCritchley, but has solved two bugs that the tests surfaced along the way.

Bug 1:
there is currently a bug in ``df.aggregate``, where the method incorrectly defers to ``df.apply`` in a corner case. This only shows up in the result when using numpy < 1.13 and passing np.nan* functions to ``df.aggregate``. This is the reason for the change in ``base.py`` line 571. (see #8383 for further details on the bug in numpy<1.13 and how it affects pandas.)

Bug 2:
Passing builtins to ``df.aggregate`` is ok when ``axis=0``, but gives wrong result,when ``axis=1`` (#21134). The reason for this difference is that ``df.aggregate`` defers to ``df._aggregate`` when ``axis=0,`` but defers to ``df.apply``, when ``axis=1``, and these give different result when passed funcions  and the series/frame contains Nan values. This can be solved by transposing df and defering the transposed frame to its ``_aggragate`` method when ``axis=1``.

The added tests have been heavily parametrized (this helped unearth the bugs above). Thet have been placed in ``series/test_apply.py`` and ``frame/test_apply``, as a lot of other tests for ser/df.aggregate were already there.

### CODE REVIEW

File: pandas/tests/test_nanops.py [hunk 1]
```diff
@@ -994,6 +994,49 @@ def prng(self):
         return np.random.RandomState(1234)
 
 
+@pytest.fixture(params=[
+    pd.Series([1, 2, 3, 4, 5, 6]),
+    pd.DataFrame([[1, 2, 3], [4, 5, 6]])
+])
+def nan_test_object(request):
```
Comment (WillAyd) [inline pandas/tests/test_nanops.py]: Should we add NA data to these?

File: pandas/core/base.py [hunk 2]
```diff
@@ -553,8 +568,8 @@ def is_any_frame():
             result = None
 
         f = self._is_cython_func(arg)
-        if f and not args and not kwargs:
-            return getattr(self, f)(), None
+        if f:
+            return getattr(self, f)(*args, **kwargs), None
```
Comment (WillAyd) [inline pandas/core/base.py]: Is this equivalent? Wondering if there's any case where providing args / kwargs before would have routed the function to a different place

File: pandas/tests/test_nanops.py [hunk 3]
```diff
@@ -994,6 +994,49 @@ def prng(self):
...
+    (np.sum, np.nansum),
+    (np.mean, np.nanmean),
+    (np.std, np.nanstd),
+    (np.var, np.nanvar),
+    (np.median, np.nanmedian),
+    (np.max, np.nanmax),
+    (np.min, np.nanmin),
+])
+def test_np_nan_functions(standard, nan_method, nan_test_object):
+    tm.assert_almost_equal(nan_test_object.agg(standard),
```
Comment (WillAyd) [inline pandas/tests/test_nanops.py]: Might be a silly question but can we not use the frame / series equals methods here? Don't think precision is that much of a factor with the fixtures

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/base.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/base.py]: It's not, but this is better :-). This fixes the bug in that was causing #19629 to fail.

The issue is subtle, but it has to do with a bug in numpy < 1.13 np.nan* functions. Numpy < 1.13 handles e.g. ``np.nanmin(pd_obj)`` incorrectly while it handles ``np.min(pd_obj)`` correctly. Numpy >= 1.13 handles both correctly. See #19753 for another issue regarding the same numpy problem.

Anyway, if ``f`` is a string, you *should* call ``getattr(self, f)(*args, **kwargs)``, so it was perhaps more luck than design that the previous version did work:-)

File: pandas/tests/test_nanops.py (hunk 3 above)
Comment (topper-123) [reply pandas/tests/test_nanops.py]: I think these are ok as ``nan_test_object`` can be either a series or a frame. I think the name should be changed though, to e.g. ``series_or_frame``.

File: pandas/tests/test_nanops.py (hunk 1 above)
Comment (topper-123) [reply pandas/tests/test_nanops.py]: Added

File: doc/source/whatsnew/v0.23.1.txt [hunk 4]
```diff
@@ -89,3 +89,9 @@ Categorical
 ^^^^^^^^^^^
 
 -
+
+Numeric
+^^^^^^^
+
+- :meth:`~DataFrame.agg` now correctly handles numpy NaN-aware methods like :meth:`numpy.nansum` (:issue:`19629`)
+- :meth:`~DataFrame.agg` now correctly handles built-in methods like ``sum`` when axis=1 (:issue:`19629`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.23.1.txt]: Don't think you meant to add this in this PR?

File: pandas/core/frame.py [hunk 5]
```diff
@@ -5818,13 +5818,12 @@ def _gotitem(self,
...
         axis = self._get_axis_number(axis)
 
-        # TODO: flipped axis
         result = None
-        if axis == 0:
-            try:
-                result, how = self._aggregate(func, axis=0, *args, **kwargs)
-            except TypeError:
-                pass
+        df = self if not axis else self.T
```
Comment (WillAyd) [inline pandas/core/frame.py]: Same thing?

File: pandas/tests/frame/test_apply.py [hunk 6]
```diff
@@ -1056,3 +1056,37 @@ def test_non_callable_aggregates(self):
...
 
         assert result == expected
+
+    @pytest.mark.parametrize("test_input", [
+        # is either frame or tuple (frame, args, kwargs),
+        pd.DataFrame([[1, 2], [3, 4]]),
+        pd.DataFrame([[np.nan, 2], [3, 4]]),
+        pd.DataFrame(),
+    ])
+    @pytest.mark.parametrize(
```
Comment (WillAyd) [inline pandas/tests/frame/test_apply.py]: Hmm wonder if this would be better as a shared fixture - thoughts?

File: pandas/tests/frame/test_apply.py [hunk 7]
```diff
@@ -1056,3 +1056,37 @@ def test_non_callable_aggregates(self):
...
+        # pd.core.base.SelectionMixin._cython_table give the same result
+        # as the related strings (values) when used in df.agg. Examples:
+        # - ``df.agg(np.nansum)`` should give the same result as
+        #   ``df.agg('sum')``
+        # - ``df.agg(sum)`` should give the same result as ``df.agg('sum')``
+        # etc.
+        # GH21123
+        np_func, str_func = cython_table_items
+
+        if isinstance(test_input, pd.DataFrame):
```
Comment (WillAyd) [inline pandas/tests/frame/test_apply.py]: Don't need this type check since the parameters are all DataFrames

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.1.txt (hunk 4 above)
Comment (topper-123) [reply doc/source/whatsnew/v0.23.1.txt]: Well, without those the tests in test_apply don't pass. At the same time, the tests in test_apply are sufficient for testing for this bug, so these two issues are very related...

File: pandas/core/frame.py (hunk 5 above)
Comment (topper-123) [reply pandas/core/frame.py]: Yeah, a bit. ``df.agg`` and ``df.apply`` with ``axis=1`` give in many cases identical output, but not in all, as ``df.apply`` doesn't make the lookup in ``_cython_table``.

AFAIK, transposition is cheap in numpy/pandas, so this is an ok approach?

File: pandas/tests/frame/test_apply.py (hunk 7 above)
Comment (topper-123) [reply pandas/tests/frame/test_apply.py]: You are right. I was preparing for eventual tests where test_input would be tuple([frame, args, kwargs]), but of course looks silly, when I could find suchs tests that make sense. 

I remove that unless someone comes up with a tests that requires args and/or kwargs

Comment (topper-123) [reply pandas/tests/frame/test_apply.py]: Also, the assert should now be ``tm.assert_frame_equal``. I'll change that.

**Comment by topper-123** [comment]
The travis failures say "Different tests were collected between gw1 and gw0. The difference is:..."

I don't *think* this has anything to do with my PR, anyone knows?

File: pandas/core/frame.py (hunk 5 above)
Comment (jreback) [reply pandas/core/frame.py]: I would rather you not do this, this is a quite hacky way of handling this, there is a small bug on the lower level i think.

File: pandas/tests/frame/test_apply.py (hunk 6 above)
Comment (jreback) [reply pandas/tests/frame/test_apply.py]: yes pls make this a fixture (in conftest.py)

File: pandas/tests/series/test_apply.py [hunk 8]
```diff
@@ -587,3 +587,34 @@ def test_map_missing_mixed(self, vals, mapping, exp):
...
         tm.assert_series_equal(result, pd.Series(exp))
+
+    @pytest.mark.parametrize("test_input", [
+        # is either series or tuple  (series, args, kwargs),
+        pd.Series([1, 2, 3, 4]),
+        pd.Series([np.nan, 2, 3, 4]),
+        pd.Series(),
+    ])
+    @pytest.mark.parametrize(
+        "cython_table_items",
```
Comment (jreback) [inline pandas/tests/series/test_apply.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 5 above)
Comment (topper-123) [reply pandas/core/frame.py]: I don't think so, as ``df._aggregate`` currently doesn't take an ``axis`` parameter. I could add an ``axis`` parameter to ``df._aggregate`` and do the transposition there instead?

Comment (jreback) [reply pandas/core/frame.py]: you can try. this should be handled on a much lower level.

**Comment by topper-123** [comment]
All the failures are of the type 

```
Different tests were collected between gw0 and gw1. The difference is:
--- gw0
+++ gw1
```

Everything passes locally. Don't think this has anything to do with my PR. I can try push again, but this has been every time so far, so don't think that will help...

**Comment by topper-123** [comment]
The failure above has been solved, it was a mistake concerning python3.5 dicts in xdist in ``conftest.py``: python3.5 randomizes dict retrieval, and of course  xdist doesn`t like randomizing work distribution. I`ve made the params parameter in ``cython_table`` deterministic, so it doesnt happen again.

File: pandas/core/frame.py (hunk 5 above)
Comment (topper-123) [reply pandas/core/frame.py]: I've added the ``axis`` parameter to ``df._aggregate``, so this is handled there.

File: pandas/core/groupby/groupby.py [hunk 9]
```diff
@@ -4086,7 +4086,10 @@ def _post_process_cython_aggregate(self, obj):
     def aggregate(self, arg, *args, **kwargs):
 
         _level = kwargs.pop('_level', None)
-        result, how = self._aggregate(arg, _level=_level, *args, **kwargs)
+        _agg_kwargs = kwargs.copy()
```
Comment (jreback) [inline pandas/core/groupby/groupby.py]: you can just list axis as a kwarg

File: pandas/conftest.py [hunk 10]
```diff
@@ -149,3 +149,20 @@ def tz_aware_fixture(request):
...
+@pytest.fixture(
+    # params: Python 3.5 randomizes dict access and xdist doesn't like that
+    # in fixtures. In order to get predetermined values we need to sort
+    # the list deterministically
+    # GH 21123
+    params=list(sorted(pd.core.base.SelectionMixin._cython_table.items(),
+                       key=lambda x: x[0].__name__)),
+    ids=lambda x: "({}-{!r})".format(x[0].__name__, x[1]),
+)
+def cython_table_items(request):
```
Comment (jreback) [inline pandas/conftest.py]: add _fixture to the end of the name

File: pandas/tests/frame/test_apply.py [hunk 11]
```diff
@@ -1056,3 +1056,26 @@ def test_non_callable_aggregates(self):
...
+        # pd.core.base.SelectionMixin._cython_table give the same result
+        # as the related strings (values) when used in df.agg. Examples:
+        # - ``df.agg(np.nansum)`` should give the same result as
+        #   ``df.agg('sum')``
+        # - ``df.agg(sum)`` should give the same result as ``df.agg('sum')``
+        # etc.
+        # GH21123
+        np_func, str_func = cython_table_items
+
+        tm.assert_almost_equal(df.agg(np_func),
```
Comment (jreback) [inline pandas/tests/frame/test_apply.py]: use assert_frame_equal it provides stronger guarantees

File: pandas/tests/series/test_apply.py [hunk 12]
```diff
@@ -587,3 +587,23 @@ def test_map_missing_mixed(self, vals, mapping, exp):
...
+        # pd.core.base.SelectionMixin._cython_table give the same result
+        # as the related strings (values), when used in ser.agg. Examples:
+        # - ``ser.agg(np.nansum)`` should give the same result as
+        #   ``ser.agg('sum')``
+        # - ``ser.agg(sum)`` should give the same result as ``ser.agg('sum')``
+        # etc.
+        # GH21123
+        np_func, str_func = cython_table_items
+
+        tm.assert_almost_equal(series.agg(np_func),
```
Comment (jreback) [inline pandas/tests/series/test_apply.py]: use assert_series_equal

File: pandas/tests/frame/test_apply.py [hunk 13]
```diff
@@ -1056,3 +1056,26 @@ def test_non_callable_aggregates(self):
...
+
+    @pytest.mark.parametrize("df", [
+        pd.DataFrame([[1, 2], [3, 4]]),
+        pd.DataFrame([[np.nan, 2], [3, 4]]),
+        pd.DataFrame(),
+    ])
+    def test_agg_function_input(self, df, cython_table_items):
+        # test whether the functions (keys) in
+        # pd.core.base.SelectionMixin._cython_table give the same result
+        # as the related strings (values) when used in df.agg. Examples:
```
Comment (jreback) [inline pandas/tests/frame/test_apply.py]: add an example which actually tests (say for sum, nansum) axis=1, IOW contruct the resultant frame

this test doesn't actually tests that axis=1 works, just that it matches with a string (which doesn't have tests itself)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_apply.py (hunk 11 above)
Comment (topper-123) [reply pandas/tests/frame/test_apply.py]: Most of these aggregate to series (so ``assert_series_equal``)), but ``cumprod`` and ``cumsum`` is in ``_cython_table``, so in that case a DataFrame is returned.

I could add a conditional, so the more correct assert is used each time.

File: pandas/conftest.py (hunk 10 above)
Comment (topper-123) [reply pandas/conftest.py]: ok

File: pandas/core/groupby/groupby.py (hunk 9 above)
Comment (topper-123) [reply pandas/core/groupby/groupby.py]: That breaks a test. The issue if ``axis`` in is in ``_kwargs``, you may get (from the breaking test):

```python
>>> _level, args, _agg_kwargs = None, (80,), {'axis': 0}
>>> self._aggregate(arg, _level=_level, *args, **_agg_kwargs)
TypeError: _aggregate() got multiple values for argument 'axis'
```

**Comment by topper-123** [comment]
I`ve rewritten the tests to now have expected results also.

File: pandas/core/base.py [hunk 14]
```diff
@@ -316,13 +331,14 @@ def _try_aggregate_string_function(self, arg, *args, **kwargs):
 
         raise ValueError("{arg} is an unknown string function".format(arg=arg))
 
-    def _aggregate(self, arg, *args, **kwargs):
+    def _aggregate(self, arg, axis=0, *args, **kwargs):
```
Comment (WillAyd) [inline pandas/core/base.py]: I still think this should be a separate PR. I know you mentioned there was some interweaving of test dependency, but I feel like we are injecting this keyword in here without any regard to existing test coverage for axis=1.

File: pandas/tests/frame/test_apply.py [hunk 15]
```diff
@@ -1056,3 +1056,66 @@ def test_non_callable_aggregates(self):
...
+        }],
+    ])
+    @pytest.mark.parametrize("axis", [0, 1], ids=lambda x: "axis {}".format(x))
+    def test_agg_function_input(self, cython_table_items, inputs, axis):
+        # GH21123
+        np_func, str_func = cython_table_items
+        df = inputs[0]
+        expected = inputs[1][str_func]
+
+        if isinstance(expected, type) and issubclass(expected, Exception):
```
Comment (WillAyd) [inline pandas/tests/frame/test_apply.py]: Anything that raises should be done in a separate test, i.e. `test_agg_function_input_raises`

File: pandas/tests/frame/test_apply.py [hunk 16]
```diff
@@ -1056,3 +1056,66 @@ def test_non_callable_aggregates(self):
...
+            with pytest.raises(expected):
+                # e.g. DataFrame(['a b'.split()]).cumprod() will raise
+                df.agg(np_func, axis=axis)
+                df.agg(str_func, axis=axis)
+        elif str_func in ('cumprod', 'cumsum'):
+            tm.assert_frame_equal(df.agg(np_func, axis=axis), expected)
```
Comment (WillAyd) [inline pandas/tests/frame/test_apply.py]: For readability / consistency with other tests create a variable called result and assign to it before the call to `assert_frame_equal`

File: pandas/tests/series/test_apply.py [hunk 17]
```diff
@@ -331,6 +331,67 @@ def test_non_callable_aggregates(self):
...
+
+        if isinstance(expected, type) and issubclass(expected, Exception):
+            with pytest.raises(expected):
+                series.agg(np_func)
+                series.agg(str_func)
+        elif str_func in ('cumprod', 'cumsum'):
+            tm.assert_series_equal(series.agg(np_func), expected)
+            tm.assert_series_equal(series.agg(str_func), expected)
+        else:
+            tm.assert_almost_equal(series.agg(np_func), expected)
```
Comment (WillAyd) [inline pandas/tests/series/test_apply.py]: Should be `assert_series_equal` no?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_apply.py (hunk 17 above)
Comment (topper-123) [reply pandas/tests/series/test_apply.py]: ``series.agg(np_func)`` resturns a scalar. I could use ``assert np.isclose(...)``?

File: pandas/tests/frame/test_apply.py (hunk 15 above)
Comment (topper-123) [reply pandas/tests/frame/test_apply.py]: I agree in principle, but this test iterates over all items in ``_cython_table``, of which some will fail on some inputs.

So I'd  have to construct the tests quite a bit differently and probably the fixture in conftest.py couldn't be used (because it returns all combinations and I now have to select the relevant ones for each test method). So something like:

```
(builtins.sum, 'sum': 0),
(np.sum, 'sum', 0),
(np.nansum: 'sum', 0),
etc...
```

which will be very inelegant and repetitive IMO. Is it not possible to bend this rule on this one (or give hint on how to do it elegantly)?...

File: pandas/core/base.py (hunk 14 above)
Comment (topper-123) [reply pandas/core/base.py]: I can look into making separate PR. That will have to be pulled in before this one, so the tests of this PR won't break.

File: pandas/tests/frame/test_apply.py (hunk 16 above)
Comment (topper-123) [reply pandas/tests/frame/test_apply.py]: I've uploaded a changed version.

File: pandas/tests/frame/test_apply.py (hunk 15 above)
Comment (WillAyd) [reply pandas/tests/frame/test_apply.py]: Hmm OK understood. May make sense as an exception then - don't have anything off the top of my head to improve but will think more about it

File: pandas/core/groupby/groupby.py (hunk 9 above)
Comment (WillAyd) [reply pandas/core/groupby/groupby.py]: Not sure I understand this - what test is breaking? Perhaps the test is configured incorrectly?

File: pandas/core/frame.py [hunk 18]
```diff
@@ -5818,13 +5818,11 @@ def _gotitem(self,
...
-        if axis == 0:
-            try:
-                result, how = self._aggregate(func, axis=0, *args, **kwargs)
-            except TypeError:
-                pass
+        try:
+            result, how = self._aggregate(func, axis=axis, *args, **kwargs)
+        except TypeError:
+            pass
         if result is None:
```
Comment (WillAyd) [inline pandas/core/frame.py]: Related to the axis change, do we still hit this condition?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py (hunk 9 above)
Comment (topper-123) [reply pandas/core/groupby/groupby.py]: Both ``80`` and ``0`` may be the value for parameter ``axis``. The signature is ``def _aggregate(self, arg, axis=0, *args, **kwargs)``, so the second unnamed argument (``80``) will be considered to be ``axis``, but this clashes with the parameter in kwargs (``{'axis': 0}``), causing the exception.

To avoid this we'd prefer the signature to be ``def _aggregate(self, arg, *args, axis=0, **kwargs)``, but this syntax is only supported in Python3...

File: pandas/core/frame.py (hunk 18 above)
Comment (topper-123) [reply pandas/core/frame.py]: The axis change is related to #21134. So if I move that to a separate PR, this will move too.

File: pandas/core/base.py (hunk 14 above)
Comment (WillAyd) [reply pandas/core/base.py]: If timing is a concern you can also xfail the axis=1 tests. Rebase thereafter would be minor

File: pandas/core/groupby/groupby.py (hunk 9 above)
Comment (WillAyd) [reply pandas/core/groupby/groupby.py]: Where is that test located?

Comment (topper-123) [reply pandas/core/groupby/groupby.py]: ``pandas\tests\groupby\test_groupby.py::test_pass_args_kwargs``

File: pandas/core/base.py (hunk 14 above)
Comment (topper-123) [reply pandas/core/base.py]: This has now been created as #21222

**Comment by jreback** [comment]
as @WillAyd indicates, can you split this up into a cython table PR and on top of that the agg fixes? 

first should be straightforward and we can get it quickly. pls put tests, changes and whatsnew for that one (in this PR is fine), and issue another PR for other changes.

File: pandas/conftest.py [hunk 19]
```diff
@@ -149,3 +149,20 @@ def tz_aware_fixture(request):
     Fixture for trying explicit timezones: {0}
     """
     return request.param
+
+
+@pytest.fixture(
+    # params: Python 3.5 randomizes dict access and xdist doesn't like that
+    # in fixtures. In order to get predetermined values we need to sort
```
Comment (jreback) [inline pandas/conftest.py]: so would like to have this fixture in your other PR

**Comment by topper-123** [comment]
Closing in favor of #22109.
